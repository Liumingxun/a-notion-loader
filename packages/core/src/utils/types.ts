import type { BlockObjectResponse, GetDataSourceResponse, PageObjectResponse, QueryDataSourceParameters, ToggleBlockObjectResponse } from '@notionhq/client/build/src/api-endpoints.d.ts'

type RecordValueOf<T> = T extends Record<string, infer U> ? U : never
type ValueOf<T> = T[keyof T]

export type PageProperties = PageObjectResponse['properties']
export type PagePropertyValue = ValueOf<PageProperties>
export type PageMetaType = Omit<PageObjectResponse, 'properties' | 'id' | 'object'> & { title: string }

export function isToggleBlock(block: BlockObjectResponse): block is ToggleBlockObjectResponse {
  return block.type === 'toggle'
}

type PropertyFilter = ([property_name, property]: [string, RecordValueOf<GetDataSourceResponse['properties']>]) => boolean
export type QueryEntriesFromDatabaseParams
  = Omit<QueryDataSourceParameters, 'filter_properties'> & { property_filter?: PropertyFilter }
