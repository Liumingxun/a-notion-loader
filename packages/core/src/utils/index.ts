import type { BlockObjectResponse } from '@notionhq/client'
import type { ToggleBlockObjectResponse } from '@notionhq/client/build/src/api-endpoints'

export { handleRichText } from './richText'
export { unescapeHTML } from 'astro/compiler-runtime'
export { escapeHTML } from 'astro/runtime/server/escape.js'

export function isToggleBlock(block: BlockObjectResponse): block is ToggleBlockObjectResponse {
  return block.type === 'toggle'
}
