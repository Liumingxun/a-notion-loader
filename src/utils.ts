import type { Client } from '@notionhq/client'
import type { Heading1BlockObjectResponse, Heading2BlockObjectResponse, Heading3BlockObjectResponse, ListBlockChildrenParameters, MentionRichTextItemResponse, PageObjectResponse, RichTextItemResponse, TableBlockObjectResponse } from '@notionhq/client/build/src/api-endpoints.d.ts'
import { isFullBlock, isFullUser } from '@notionhq/client'
import { escapeHTML } from 'astro/runtime/server/escape.js'

function applyAnnotations(content: string, annotations: Record<string, boolean>): string {
  const annotationMap: Record<string, (text: string) => string> = {
    bold: text => `<strong>${text}</strong>`,
    italic: text => `<em>${text}</em>`,
    strikethrough: text => `<s>${text}</s>`,
    underline: text => `<u>${text}</u>`,
    code: text => `<code>${text}</code>`,
  }

  return Object.keys(annotations)
    .filter(key => annotations[key])
    .reduce((result, key) => annotationMap[key]?.(result) || result, content)
}

function handleMention(mentionBlock: MentionRichTextItemResponse): string {
  const { mention } = mentionBlock
  if (!mention)
    return ''

  if (mention.type === 'user' && isFullUser(mention.user)) {
    const user = mention.user
    if (user.type === 'person') {
      return `<a href="mailto:${user.person.email}"><span><img style="width: 1em; display: inline-block" src="${user.avatar_url}" alt="${mentionBlock.plain_text}'s avatar" />${mentionBlock.plain_text}</span></a>`
    }
  }
  else if (mention.type === 'date') {
    const start = `<time datetime="${mention.date.start}">${mention.date.start}</time>`
    const end = mention.date.end ? ` ~ <time datetime="${mention.date.end}">${mention.date.end}</time>` : ''
    return start + end
  }
  else if (mention.type === 'page') {
    return `<a href="/${mention.page.id}">${mentionBlock.plain_text}</a>` // TODO: prefix with site URL according to config
  }

  return ''
}

export function handleRichText(richTextList: Array<RichTextItemResponse> | undefined, plain: boolean = false): string {
  if (!richTextList || richTextList.length === 0)
    return ''
  if (plain)
    return richTextList.reduce((frag, item) => frag + item.plain_text, '')
  return richTextList.reduce((frag, item) => {
    let content = escapeHTML(item.plain_text)

    if (item.type === 'text' && item.href) {
      content = `<a href="${item.href}">${content}</a>`
    }
    else if (item.type === 'mention') {
      content = handleMention(item) || content
    }

    const { color, ...filteredAnnotations } = item.annotations
    if (!['default', 'default_background'].includes(color)) {
      const style = color.endsWith('background') ? `background-color: ${color.replace('_background', '')}` : `color: ${color}`
      content = `<mark style="${style}">${content}</mark>`
    }

    content = applyAnnotations(content, filteredAnnotations)
    return frag + content
  }, '')
}

async function handleHeading(headingBlock: Heading1BlockObjectResponse | Heading2BlockObjectResponse | Heading3BlockObjectResponse, client: Client): Promise<string> {
  const { type } = headingBlock
  let heading: string = ''
  if (type === 'heading_1') {
    heading = `<h1>${handleRichText(headingBlock.heading_1.rich_text)}</h1>`
  }
  else if (type === 'heading_2') {
    heading = `<h2>${handleRichText(headingBlock.heading_2.rich_text)}</h2>`
  }
  else if (type === 'heading_3') {
    heading = `<h3>${handleRichText(headingBlock.heading_3.rich_text)}</h3>`
  }
  if (headingBlock.has_children) {
    const { content: headingContent } = await handleChildren({ block_id: headingBlock.id }, client)
    heading = `<details><summary>${heading}</summary>${headingContent}</details>`
  }
  return heading
}

async function handleTable(tableBlock: TableBlockObjectResponse, client: Client) {
  const { table: { has_column_header, has_row_header } } = tableBlock
  const { results } = await client.blocks.children.list({ block_id: tableBlock.id })

  const rows = results.filter(r => isFullBlock(r) && r.type === 'table_row').map(r => r.table_row.cells)

  const tableRows = rows.map((row, rowIndex) => {
    const cells = row.map((cell, colIndex) => {
      const cellContent = handleRichText(cell)
      if (has_row_header && rowIndex === 0) {
        return `<th scope="col">${cellContent}</th>`
      }
      if (has_column_header && colIndex === 0) {
        return `<th scope="row">${cellContent}</th>`
      }
      return `<td>${cellContent}</td>`
    }).join('')

    return `<tr>${cells}</tr>`
  }).join('')

  return `<table>${tableRows}</table>`
}

export async function handleChildren(query: ListBlockChildrenParameters, client: Client) {
  const { results } = await client.blocks.children.list(query)
  let content = ''
  for (const block of results.filter(r => isFullBlock(r))) {
    if (block.type === 'paragraph') {
      content += `<p>${handleRichText(block.paragraph.rich_text)}</p>`
    }
    else if (block.type === 'divider') {
      content += '<hr />'
    }
    else if (block.type === 'code') {
      content += `\n\n\`\`\`${block.code.language}\n${handleRichText(block.code.rich_text)}\n\`\`\`\n\n`
    }
    else if (block.type === 'heading_1' || block.type === 'heading_2' || block.type === 'heading_3') {
      content += await handleHeading(block, client)
    }
    else if (block.type === 'table') {
      content += await handleTable(block, client)
    }
  }
  return { content }
}

type PageProperties = PageObjectResponse['properties']
type ValueOf<T> = T[keyof T]
type PagePropertyValue = ValueOf<PageProperties>

export type PropertiesType<T extends PagePropertyValue = PagePropertyValue> = Array<{ label: string, value: T }>
export type MetaType = Partial<Omit<PageObjectResponse, 'properties' | 'id' | 'object'> & {
  title: string
}>
