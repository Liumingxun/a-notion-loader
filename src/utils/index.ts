export { handleProperty } from './property'
export { handleRichText } from './richText'
export * from './types'
export { unescapeHTML } from 'astro/compiler-runtime'
export { escapeHTML } from 'astro/runtime/server/escape.js'
