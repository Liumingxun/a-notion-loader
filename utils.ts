import type { ParagraphBlockObjectResponse } from '@notionhq/client/build/src/api-endpoints'

export function reduceParagraph(cur: ParagraphBlockObjectResponse) {
  return cur.paragraph.rich_text.reduce((mdText, item) => {
    if (item.type === 'text') {
      let content = item.plain_text
      const href = item.href
      const color = item.annotations.color === 'default' ? '' : item.annotations.color
      const annotations = (Object.keys(item.annotations) as ('bold' | 'italic' | 'strikethrough' | 'underline' | 'code')[])
        .filter(v => typeof item.annotations[v] === 'boolean' && item.annotations[v] === true)
      if (href) {
        content = `<a href="${href}">${content}</a>`
      }
      if (color) {
        content = `<mark color="${color}">${content}</mark>`
      }
      if (annotations.length) {
        content = annotations.reduce((mark, annotation) => {
          if (annotation === 'bold') {
            mark = `<strong>${mark}</strong>`
          }
          else if (annotation === 'italic') {
            mark = `<em>${mark}</em>`
          }
          else if (annotation === 'strikethrough') {
            mark = `<s>${mark}</s>`
          }
          else if (annotation === 'underline') {
            mark = `<u>${mark}</u>`
          }
          else if (annotation === 'code') {
            mark = `<code>${mark}</code>`
          }
          return mark
        }, content)
      }
      mdText += content
    }
    else if (item.type === 'mention') {
      let content = item.plain_text

      if (item.mention.type === 'date') {
        content = item.mention.date.start
      }

      mdText += content
    }

    return mdText
  }, '')
}
