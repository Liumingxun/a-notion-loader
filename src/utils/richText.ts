import type { MentionRichTextItemResponse, RichTextItemResponse, RichTextItemResponseCommon } from '@notionhq/client/build/src/api-endpoints.d.ts'
import { isFullUser } from '@notionhq/client'
import { isMentionRichTextItemResponse } from '@notionhq/client/build/src/helpers'
import { escapeHTML } from '.'

function applyAnnotations(content: string, annotations: Record<string, boolean>): string {
  const annotationMap: Record<string, (text: string) => string> = {
    bold: text => `<strong>${text}</strong>`,
    italic: text => `<em>${text}</em>`,
    strikethrough: text => `<s>${text}</s>`,
    underline: text => `<u>${text}</u>`,
    code: text => `<code>${text}</code>`,
  }

  return Object.keys(annotations)
    .filter(key => annotations[key])
    .reduce((result, key) => annotationMap[key]?.(result) || result, content)
}

function handleMention(mentionBlock: RichTextItemResponseCommon & MentionRichTextItemResponse): string {
  const { mention } = mentionBlock

  if (mention.type === 'user' && isFullUser(mention.user)) {
    const user = mention.user
    if (user.type === 'person') {
      return `<a href="mailto:${user.person.email}"><span><img style="width: 1em; display: inline-block" src="${user.avatar_url}" alt="${mentionBlock.plain_text}'s avatar" />${mentionBlock.plain_text}</span></a>`
    }
  }
  else if (mention.type === 'date') {
    const start = `<time datetime="${mention.date.start}">${mention.date.start}</time>`
    const end = mention.date.end ? ` ~ <time datetime="${mention.date.end}">${mention.date.end}</time>` : ''
    return start + end
  }
  else if (mention.type === 'page') {
    return `<a href="/${mention.page.id}">${mentionBlock.plain_text}</a>` // TODO: prefix with site URL according to config
  }
  else if (mention.type === 'link_mention') {
    const { link_mention: { href, link_provider, title, icon_url } } = mention
    return `<a href="${href}"><span><img style="width: 1em; display: inline-block" src="${icon_url}" ${link_provider ? `alt="${link_provider}'s favicon"` : ''}  />${title}</a>`
  }

  return ''
}

export function handleRichText(richTextList: Array<RichTextItemResponse> | undefined, plain: boolean = false): string {
  if (!richTextList || richTextList.length === 0)
    return ''
  if (plain)
    return richTextList.reduce((frag, item) => frag + item.plain_text, '')
  return richTextList.reduce((frag, item) => {
    let content = escapeHTML(item.plain_text)

    if (isMentionRichTextItemResponse(item)) {
      content = handleMention(item) || content
    }
    else if (item.type === 'text') {
      if (item.href)
        content = `<a href="${item.href}">${content}</a>`
    }

    const { color, ...filteredAnnotations } = item.annotations
    if (!['default', 'default_background'].includes(color)) {
      const style = color.endsWith('background') ? `background-color: ${color.replace('_background', '')}` : `color: ${color}`
      content = `<mark style="${style}">${content}</mark>`
    }

    content = applyAnnotations(content, filteredAnnotations)
    return frag + content
  }, '')
}
