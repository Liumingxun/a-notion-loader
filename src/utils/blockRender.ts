import type { Client } from '@notionhq/client'
import type {
  BlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ColumnListBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ListBlockChildrenResponse,
  NumberedListItemBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  TableBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints.d.ts'
import type { LoaderContext } from 'astro/loaders'
import { isFullBlock } from '@notionhq/client'
import { unescapeHTML } from 'astro/compiler-runtime'
import { handleRichText } from './richText'
import { isListItemBlock, isToggleBlock } from './types'

export class NotionRenderer {
  constructor(
    public client: Client,
    public renderMarkdown: LoaderContext['renderMarkdown'],
  ) {}

  async fetchAllChildren(id: string) {
    const allResults: ListBlockChildrenResponse['results'] = []
    let cursor: string | undefined

    while (true) {
      const response = await this.client.blocks.children.list({
        block_id: id,
        start_cursor: cursor,
      })

      allResults.push(...response.results)

      if (!response.has_more)
        break
      cursor = response.next_cursor ?? undefined
    }

    return allResults
  }

  async handleChildren(allChildren: ListBlockChildrenResponse['results']) {
    const content: string[] = []
    const blocks = allChildren.filter(r => isFullBlock(r))

    for (let i = 0; i < blocks.length; i++) {
      const block = blocks[i]

      if (block.type === 'paragraph') {
        content.push(`<p>${handleRichText(block.paragraph.rich_text)}</p>`)
      }
      else if (block.type === 'divider') {
        content.push('<hr />')
      }
      else if (block.type === 'code') {
        content.push((await this.renderMarkdown(`\n\n\`\`\`${block.code.language}\n${unescapeHTML(handleRichText(block.code.rich_text, true))}\n\`\`\`\n\n`)).html)
      }
      else if (block.type === 'heading_1' || block.type === 'heading_2' || block.type === 'heading_3') {
        content.push(await this.handleHeading(block))
      }
      else if (block.type === 'table') {
        content.push(await this.handleTable(block))
      }
      else if (block.type === 'callout') {
        content.push(this.handleCallout(block))
      }
      else if (block.type === 'column_list') {
        content.push(await this.handleColumnList(block))
      }
      else if (block.type === 'to_do') {
        content.push(await this.handleTodo(block))
      }
      else if (block.type === 'quote') {
        content.push(await this.handleQuote(block))
      }
      else if (block.type === 'toggle') {
        content.push(await this.handleToggle(block))
      }
      else if (block.type === 'bulleted_list_item' || block.type === 'numbered_list_item') {
        const { result, skipCount } = await this.handleList(blocks.slice(i))
        i += skipCount
        content.push(result)
      }
    }

    return { content: content.join('\n\n') }
  }

  async renderAllChildren(id: string) {
    const allChildren = await this.fetchAllChildren(id)
    return this.handleChildren(allChildren)
  }

  async handleHeading(headingBlock: Heading1BlockObjectResponse | Heading2BlockObjectResponse | Heading3BlockObjectResponse) {
    const { type } = headingBlock
    let heading: string = ''
    let isToggleable = false
    if (type === 'heading_1') {
      heading = `# ${handleRichText(headingBlock.heading_1.rich_text)}`
      isToggleable = headingBlock.heading_1.is_toggleable
    }
    else if (type === 'heading_2') {
      heading = `## ${handleRichText(headingBlock.heading_2.rich_text)}`
      isToggleable = headingBlock.heading_2.is_toggleable
    }
    else if (type === 'heading_3') {
      heading = `### ${handleRichText(headingBlock.heading_3.rich_text)}`
      isToggleable = headingBlock.heading_3.is_toggleable
    }
    if (!isToggleable)
      return heading

    return await this.handleToggle(headingBlock)
  }

  async handleTable(tableBlock: TableBlockObjectResponse) {
    const { table: { has_column_header, has_row_header }, id } = tableBlock
    const { results } = await this.client.blocks.children.list({ block_id: id })

    const rows = results.filter(r => isFullBlock(r) && r.type === 'table_row').map(r => r.table_row.cells)

    const tableRows = rows.map((row, rowIndex) => {
      const cells = row.map((cell, colIndex) => {
        const cellContent = handleRichText(cell)
        if (has_row_header && rowIndex === 0) {
          return `<th scope="col">${cellContent}</th>`
        }
        if (has_column_header && colIndex === 0) {
          return `<th scope="row">${cellContent}</th>`
        }
        return `<td>${cellContent}</td>`
      }).join('')

      return `<tr>${cells}</tr>`
    }).join('')

    return `<table>${tableRows}</table>`
  }

  handleCallout(calloutBlock: CalloutBlockObjectResponse): string {
    const { callout: { icon, rich_text } } = calloutBlock

    let iconHTML = ''
    if (icon) {
      if (icon.type === 'emoji') {
        iconHTML = `<span>${icon.emoji}</span>`
      }
      else if (icon.type === 'external') {
        iconHTML = `<img style="width: 1.25rem; display: inline-block" src="${icon.external.url}" alt="Icon" />`
      }
      else if (icon.type === 'custom_emoji') {
        iconHTML = `<img style="width: 1.25rem; display: inline-block" src="${icon.custom_emoji.url}" alt="${icon.custom_emoji.name}" />`
      }
      else if (icon.type === 'file') {
        iconHTML = `<img style="width: 1.25rem; display: inline-block" src="${icon.file.url}" alt="Icon" />`
      }
    }

    const text = `<span>${handleRichText(rich_text)}</span>`
    return `<div style="display: flex; padding: 0.5rem; align-items: baseline; gap: 0.25rem;">${iconHTML}<div>${text}</div></div>`
  }

  async handleListItem(listItemBlock: BulletedListItemBlockObjectResponse | NumberedListItemBlockObjectResponse) {
    const { type, id, has_children } = listItemBlock
    const { rich_text } = type === 'bulleted_list_item' ? listItemBlock.bulleted_list_item : listItemBlock.numbered_list_item

    const content = `<span>${handleRichText(rich_text)}</span>`

    if (!has_children) {
      return `<li>${content}</li>`
    }

    const { content: childContent } = await this.renderAllChildren(id)
    return `<li>${content}${childContent}</li>`
  }

  async handleList(blocks: BlockObjectResponse[]) {
    const firstBlock = blocks[0] as BulletedListItemBlockObjectResponse | NumberedListItemBlockObjectResponse
    const listItems = [firstBlock]
    const listType = firstBlock.type

    let count = 1
    while (count < blocks.length) {
      const b = blocks[count]
      if (isListItemBlock(b) && b.type === listType) {
        listItems.push(b)
        count++
      }
      else {
        break
      }
    }

    const listHtml = await Promise.all(listItems.map(item => this.handleListItem(item)))

    const wrapperTag = listType === 'bulleted_list_item' ? 'ul' : 'ol'

    const result = `<${wrapperTag}>\n${listHtml.join('\n')}\n</${wrapperTag}>`
    return { result, skipCount: count - 1 }
  }

  async handleColumnList(columnListBlock: ColumnListBlockObjectResponse) {
    const { results } = await this.client.blocks.children.list({ block_id: columnListBlock.id })
    const columns = results.filter(r => isFullBlock(r) && r.type === 'column')
      .map(async (column) => {
        return `<div>${(await this.renderAllChildren(column.id)).content}</div>`
      })

    return `<div style="display: flex; gap: 1rem;">${await Promise.all(columns).then(cols => cols.join(''))}</div>`
  }

  async handleTodo(todoBlock: ToDoBlockObjectResponse) {
    const { id, has_children, to_do: { rich_text, checked } } = todoBlock
    const checkbox = `<input disabled type="checkbox" ${checked ? 'checked' : ''} />`
    const textContent = `<span>${handleRichText(rich_text)}</span>`

    if (!has_children) {
      return `<div style="display: flex; gap: 0.5rem;">${checkbox}${textContent}</div>`
    }

    const { content: childContent } = await this.renderAllChildren(id)
    return `<div style="display: flex; gap: 0.5rem; align-items: baseline">${checkbox}<div>${textContent}${childContent}</div></div>`
  }

  async handleQuote(quoteBlock: QuoteBlockObjectResponse) {
    const { quote: { rich_text }, has_children } = quoteBlock
    const content = `<span>${handleRichText(rich_text)}</span>`

    if (!has_children)
      return `<blockquote>${content}</blockquote>`

    const { content: childContent } = await this.renderAllChildren(quoteBlock.id)
    return `<blockquote >${content}${childContent}</blockquote>`
  }

  async handleToggle(toggleBlock: ToggleBlockObjectResponse | Heading1BlockObjectResponse | Heading2BlockObjectResponse | Heading3BlockObjectResponse) {
    const { has_children } = toggleBlock
    let richText: RichTextItemResponse[], tag: string
    if (isToggleBlock(toggleBlock)) {
      richText = toggleBlock.toggle.rich_text
      tag = 'span'
    }
    else {
      const headingInfo = getHeadingInfo(toggleBlock)
      richText = headingInfo.richText
      tag = headingInfo.tag
    }

    const content = `<${tag}>${handleRichText(richText)}</${tag}>`
    if (!has_children)
      return `<details><summary>${content}</summary></details>`

    const { content: childContent } = await this.renderAllChildren(toggleBlock.id)
    return `<details><summary>${content}</summary>${childContent}</details>`
  }
}

function getHeadingInfo(block: Heading1BlockObjectResponse | Heading2BlockObjectResponse | Heading3BlockObjectResponse) {
  if (block.type === 'heading_1') {
    return {
      tag: 'h1' as const,
      richText: block.heading_1.rich_text,
    }
  }
  if (block.type === 'heading_2') {
    return {
      tag: 'h2' as const,
      richText: block.heading_2.rich_text,
    }
  }

  return {
    tag: 'h3' as const,
    richText: block.heading_3.rich_text,
  }
}
