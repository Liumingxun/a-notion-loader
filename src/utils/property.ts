import type { Client } from '@notionhq/client'
import type { GetPagePropertyParameters, PeoplePropertyItemObjectResponse, PropertyItemObjectResponse, RelationPropertyItemObjectResponse } from '@notionhq/client/build/src/api-endpoints.d.ts'
import { handleRichText } from './richText'

export async function handleProperty(query: GetPagePropertyParameters, client: Client) {
  // TODO: shounldn't query properties in this function
  const property = await client.pages.properties.retrieve(query)

  if (property.object === 'list') {
    const { results, property_item: { type } } = property
    if (type === 'title') {
      return handleRichText(results.filter(p => p.type === 'title').map(p => p.title), true)
    }
    else if (type === 'rich_text') {
      return handleRichText(results.filter(p => p.type === 'rich_text').map(p => p.rich_text))
    }
    else if (type === 'relation') {
      return results as RelationPropertyItemObjectResponse[]
    }
    else if (type === 'people') {
      return results as PeoplePropertyItemObjectResponse[]
    }
    else if (type === 'rollup' && property.property_item.type === 'rollup') {
      const { rollup } = property.property_item
      return { rollup, results }
    }
  }
  return property as PropertyItemObjectResponse
}
