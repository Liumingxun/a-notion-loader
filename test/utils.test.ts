import type { ListBlockChildrenResponse, PageObjectResponse, ParagraphBlockObjectResponse } from '@notionhq/client/build/src/api-endpoints'
import { isFullBlock } from '@notionhq/client'
import { expect, it } from 'vitest'
import { reduceRichText } from '../utils'
import content from './paragraph.json'

(content as ListBlockChildrenResponse).results.filter(r => isFullBlock(r) && r.type === 'paragraph').map(p => it(`paragraph`, () => {
  return expect(reduceRichText(p.paragraph.rich_text)).toMatchSnapshot()
}))
