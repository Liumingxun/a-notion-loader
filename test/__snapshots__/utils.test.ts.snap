// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`paragraph 1`] = `"notag"`;

exports[`paragraph 2`] = `"<strong>bold</strong>"`;

exports[`paragraph 3`] = `"<em>italic</em>"`;

exports[`paragraph 4`] = `"<s>strikethrough</s>"`;

exports[`paragraph 5`] = `"<u>underline</u>"`;

exports[`paragraph 6`] = `"<code>code</code>"`;

exports[`paragraph 7`] = `"<em><strong>bold and italic</strong></em>"`;

exports[`paragraph 8`] = `"<s><em><strong>bold and italic and strikethrough</strong></em></s>"`;

exports[`paragraph 9`] = `"<u><s><em><strong>bold and italic and strikethrough and underline</strong></em></s></u>"`;

exports[`paragraph 10`] = `"<code><u><s><em><strong>bold and italic and strikethrough and underline and code</strong></em></s></u></code>"`;

exports[`paragraph 11`] = `"<code><u><s><em><strong><a href="https://notion.so/">bold and italic and strikethrough and underline and code and link</a></strong></em></s></u></code>"`;

exports[`paragraph 12`] = `"notag with <strong>bold,</strong> <em>italic, </em><s>strikethrough, </s><u>underline, </u><code>code</code>, <em><strong>bold and italic, </strong></em><s><em><strong>bold and italic and strikethrough, </strong></em></s><u><s><em><strong>bold and italic and strikethrough and underline, </strong></em></s></u><code><u><s><em><strong>bold and italic and strikethrough and underline and code</strong></em></s></u></code><u><s><em><strong> , </strong></em></s></u><code><u><s><em><strong><a href="https://notion.so/">bold and italic and strikethrough and underline and code and link</a></strong></em></s></u></code>"`;

exports[`paragraph 13`] = `"default"`;

exports[`paragraph 14`] = `"<mark color="gray">hello gray</mark>"`;

exports[`paragraph 15`] = `"<mark color="brown">hello brown</mark>"`;

exports[`paragraph 16`] = `"<mark color="orange_background">hello orange background</mark>"`;

exports[`paragraph 17`] = `"<mark color="yellow_background">hello yellow background</mark>"`;

exports[`paragraph 18`] = `"<mark color="green">hello green and blue background</mark>"`;

exports[`paragraph 19`] = `"<mark color="purple">hello purple and pink background</mark>"`;

exports[`paragraph 20`] = `"default, <mark color="gray">hello gray, </mark><mark color="brown">hello brown, </mark><mark color="orange_background">hello orange background, </mark><mark color="yellow_background">hello yellow background, </mark><mark color="green">hello green and blue background, </mark><mark color="purple">hello purple and pink background</mark>"`;

exports[`paragraph 21`] = `"<a href="https://notion.so/">default, </a><mark color="gray"><a href="https://notion.so/">hello gray, </a></mark><mark color="brown"><a href="https://notion.so/">hello brown, </a></mark><mark color="orange_background"><a href="https://notion.so/">hello orange background, </a></mark><mark color="yellow_background"><a href="https://notion.so/">hello yellow background, </a></mark><mark color="green"><a href="https://notion.so/">hello green and blue background, </a></mark><mark color="purple"><a href="https://notion.so/">hello purple and pink background</a></mark>"`;

exports[`paragraph 22`] = `"<strong><mark color="green">bold and green</mark></strong>"`;

exports[`paragraph 23`] = `"<em><mark color="yellow_background">italic and yellow background</mark></em>"`;

exports[`paragraph 24`] = `"<s><mark color="blue">strikethrough and blue and purple background</mark></s>"`;

exports[`paragraph 25`] = `"<strong><mark color="green">bold and green, </mark></strong><em><mark color="yellow_background">italic and yellow background, </mark></em><s><mark color="blue">strikethrough and blue and purple background</mark></s>"`;

exports[`paragraph 26`] = `"<strong><mark color="green">bold and green, </mark></strong><em><mark color="yellow_background">italic and yellow background, </mark></em><s><mark color="blue">strikethrough and blue and purple background</mark></s> but block color"`;

exports[`paragraph 27`] = `"@Luoming"`;

exports[`paragraph 28`] = `"2025-05-05 "`;

exports[`paragraph 29`] = `"2025-05-06T09:00:00.000+08:00 "`;

exports[`paragraph 30`] = `"sub page "`;

exports[`paragraph 31`] = `"<em><strong>bold and italic </strong></em>@Luoming<em><strong> </strong></em>"`;

exports[`paragraph 32`] = `"<mark color="brown">hello brown </mark>2025-05-05<mark color="brown"> </mark>"`;

exports[`paragraph 33`] = `"<mark color="green">hello green and blue background </mark>sub page<mark color="green"> </mark>"`;

exports[`paragraph 34`] = `"2025-05-05 <s><mark color="blue">strikethrough and blue and purple background </mark></s>notag with <strong>bold,</strong> <em>italic, </em><s>strikethrough, </s><u>underline, </u><code>code</code>, <em><strong>bold and italic </strong></em>@Luoming<em><strong> </strong></em>"`;

exports[`paragraph 35`] = `"paragraph "`;

exports[`paragraph 36`] = `"https://developers.notion.com/reference/rich-text#equation"`;

exports[`paragraph 37`] = `""`;
